src/domains.cpp

```diff
@@ -74,6 +77,145 @@ void domain_build (domain_param *domain_par,
+                                  dolfin::Function *anion,
+                                  dolfin::Function *voltage,
+                                  dolfin::Mesh *target_mesh,
+                                  double entropy_tol)
+{
+  // compute mesh from input voltage function and transfer
+  dolfin::Mesh mesh( *(voltage->function_space()->mesh()) );
+  gradient_recovery::FunctionSpace gradient_space(mesh);
+  gradient_recovery::BilinearForm a(gradient_space,gradient_space);
+  gradient_recovery::LinearForm L(gradient_space);
+  dolfin::Function cation_entropy(gradient_space);
+  dolfin::Function anion_entropy(gradient_space);
+
+  // compute entropic potentials
+  dolfin::Function cation_potential( *(voltage->function_space()) );
+  cation_potential.interpolate(*voltage);
+  *(cation_potential.vector()) *= +1.0;
+  *(cation_potential.vector()) += *(cation->vector());
+  dolfin::Function anion_potential( *(voltage->function_space()) );
+  anion_potential.interpolate(*voltage);
+  *(anion_potential.vector()) *= -1.0;
+  *(anion_potential.vector()) += *(anion->vector());
+
+  // setup matrix and rhs
+  EigenMatrix A;
+  assemble(A,a);
+  dCSRmat A_fasp;
+  EigenMatrix_to_dCSRmat(&A,&A_fasp);
+  dBSRmat adaptA_fasp_bsr = fasp_format_dcsr_dbsr(&A_fasp, 3);
+  EigenVector b;
+  dvector b_fasp, solu_fasp;
+
+  // Setup FASP solver
+  input_param inpar;
+  itsolver_param itpar;
+  AMG_param amgpar;
+  ILU_param ilupar;
+  char fasp_param_file[] = "./src/gradient_recovery_bsr.dat";
+  fasp_param_input(fasp_param_file, &inpar);
+  fasp_param_init(&inpar, &itpar, &amgpar, &ilupar, NULL);
+  INT status = FASP_SUCCESS;
+
+  // set form for cation
+  L.potential = cation_potential;
+  assemble(b,L);
+  EigenVector_to_dvector(&b,&b_fasp);
+  fasp_dvec_alloc(b.size(), &solu_fasp);
+
+  // solve for cation entropy
+  fasp_dvec_set(b_fasp.row, &solu_fasp, 0.0);
+  status = fasp_solver_dbsr_krylov_diag(&adaptA_fasp_bsr, &b_fasp, &solu_fasp, &itpar);
+  copy_dvector_to_Function(&solu_fasp, &cation_entropy);
+
+  // set form for anion
+  L.potential = anion_potential;
+  assemble(b,L);
+  EigenVector_to_dvector(&b,&b_fasp);
+
+  // solve for anion entropy
+  fasp_dvec_set(b_fasp.row, &solu_fasp, 0.0);
+  status = fasp_solver_dbsr_krylov_diag(&adaptA_fasp_bsr, &b_fasp, &solu_fasp, &itpar);
+  copy_dvector_to_Function(&solu_fasp, &anion_entropy);
+
+  // output entropy
+  // File entropyFile("./benchmarks/PNP/output/entropy.pvd");
+  // entropyFile << cation_entropy;
+  // entropyFile << anion_entropy;
+
+  // compute entropic error
+  poisson_cell_marker::FunctionSpace DG(mesh);
+  poisson_cell_marker::LinearForm error_form(DG);
+  error_form.cat_entr = cation_entropy;
+  error_form.cat_pot  = cation_potential;
+  error_form.an_entr = anion_entropy;
+  error_form.an_pot  = anion_potential;
+  dolfin::EigenVector error_vector;
+  assemble(error_vector, error_form);
+
+  // mark for refinement
+  MeshFunction<bool> cell_marker(mesh, 3, false);
```

Comment:
Instead of 3 you should put "mesh.topology().dim()" to make it more general ? Because we will to some 2D.